File: llmc/orchestrator.py
"""Docker SDK wrapper for the v2 proxy.

Replaces the legacy `docker compose --profile X up/stop` shellout pattern
with direct calls to the Docker Engine API via the `docker` Python SDK.
This eliminates:

    - .env file rewriting (env vars are passed directly to containers.run)
    - compose profile dance (each GPU service is just a container we run/stop)
    - the docker-compose CLI dependency in the proxy image
    - shell-parsing of compose error output

The orchestrator manages three mutually-exclusive GPU services
(llama-server, comfyui, lora-train). Only one runs at a time — the
proxy invokes `stop_gpu_services()` before spawning a new one.

Containers are tagged with two labels so the orchestrator can find them
again after a proxy restart:

    llmc.service = "llama-server" | "comfyui" | "lora-train"
    llmc.mode    = "llm" | "comfyui" | "train"

`current_mode()` reads those labels off running containers and returns
the active mode (or "idle" if none are running).
"""

from __future__ import annotations

import http.client
import io
import os
import tarfile
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Optional, TYPE_CHECKING

# The `docker` package is the proxy's only runtime dep, but the rest of
# the llmc package (CLI, presets, volumes, state) is stdlib-only. Import
# docker lazily so test modules that don't construct an Orchestrator can
# still be imported on a host without the SDK installed.
if TYPE_CHECKING:
    import docker
    from docker.models.containers import Container

from llmc.presets import Preset, preset_to_env
from llmc.volumes import VolumeRegistry


# ── Service definitions ────────────────────────────────────────────────


@dataclass(frozen=True)
class GpuService:
    name: str           # Container name (matches `docker ps`)
    hostname: str       # Hostname inside the user-defined network
    mode: str           # "llm" | "comfyui" | "train"
    image: str
    internal_port: int  # Port the service listens on inside the container
    health_path: str    # Path on the service to poll for readiness


LLAMA_SERVICE = GpuService(
    name="llama_server",
    hostname="llama-server",
    mode="llm",
    image=os.environ.get("LLMC_LLAMA_IMAGE", "erfianugrah/llama-server:cuda12.8-sm120"),
    internal_port=8080,
    health_path="/health",
)

# NInfer serves the same mode as llama.cpp: both are the LLM, and the GPU
# holds one workload at a time. It is deliberately NOT in SERVICES (which
# maps mode -> service 1:1); resolve it from the preset via llm_service_for.
NINFER_SERVICE = GpuService(
    name="ninfer_server",
    hostname="ninfer-server",
    mode="llm",
    image=os.environ.get("LLMC_NINFER_IMAGE", "erfianugrah/ninfer:cuda13.1-sm120a-d492968"),
    internal_port=8080,
    health_path="/health",
)

COMFYUI_SERVICE = GpuService(
    name="comfyui",
    hostname="comfyui",
    mode="comfyui",
    image=os.environ.get("LLMC_COMFYUI_IMAGE", "erfianugrah/comfyui:cuda12.8-sm120"),
    internal_port=8188,
    health_path="/system_stats",
)

TRAIN_SERVICE = GpuService(
    name="lora_train",
    hostname="lora-train",
    mode="train",
    image=os.environ.get("LLMC_TRAIN_IMAGE", "erfianugrah/lora-train:latest"),
    internal_port=8787,
    health_path="/health",
)

SERVICES = {svc.mode: svc for svc in (LLAMA_SERVICE, COMFYUI_SERVICE, TRAIN_SERVICE)}

GPU_LABEL = "llmc.mode"
SERVICE_LABEL = "llmc.service"
DEFAULT_NETWORK = os.environ.get("LLMC_NETWORK", "llmc")
DEFAULT_HEALTH_TIMEOUT = int(os.environ.get("LLMC_HEALTH_TIMEOUT", "900"))


def llm_service_for(preset: Preset) -> GpuService:
    """Which container serves this preset. Two engines share mode "llm", so
    the mode alone cannot decide - the preset does."""
    if preset.engine == "ninfer":
        return NINFER_SERVICE
    return LLAMA_SERVICE


def ninfer_command(preset: Preset) -> list[str]:
    """Build the `ninfer-serve` argv for a ninfer preset.

    Unlike the llama image (whose ENTRYPOINT assembles a command line from
    environment variables), ninfer-serve is driven directly by argv, so the
    proxy owns the flag rendering. Returned as a list because the Docker SDK
    shlex-splits string commands at whitespace and mangles them.
    """
    if preset.engine != "ninfer":
        raise ValueError(
            f"ninfer_command: {preset.name!r} declares engine={preset.engine!r}"
        )
    spec = preset.ninfer
    argv: list[str] = [
        "ninfer-serve",
        f"/models/{preset.model.file}",
        "--model-id", preset.model_id,
        "--host", "0.0.0.0",
        "--port", str(NINFER_SERVICE.internal_port),
        "--max-context", str(spec.max_context),
        "--max-concurrency", str(spec.max_concurrency),
        "--kv-dtype", spec.kv_dtype,
    ]
    # `is not None` rather than truthiness: 0 is a meaningful value for the
    # host/device slot flags (the WSL2 pinned-host workaround).
    for flag, value in (
        ("--spec", spec.spec),
        ("--draft-tokens", spec.draft_tokens),
        ("--host-state-slots", spec.host_state_slots),
        ("--host-kv-mib", spec.host_kv_mib),
        ("--device-state-slots", spec.device_state_slots),
        ("--default-thinking-budget", spec.default_thinking_budget),
        ("--kv-capacity", spec.kv_capacity),
        ("--prefill-chunk", spec.prefill_chunk),
        ("--max-pending-requests", spec.max_pending_requests),
        ("--pending-timeout-ms", spec.pending_timeout_ms),
        ("--request-log-jsonl", spec.request_log_jsonl),
    ):
        if value is not None:
            argv += [flag, str(value)]
    for flag, enabled in (
        ("--lm-head-draft", spec.lm_head_draft),
        ("--preserve-thinking", spec.preserve_thinking),
        ("--vision", spec.vision),
    ):
        if enabled:
            argv.append(flag)
    return argv


class OrchestratorError(RuntimeError):
    """Raised when a Docker operation fails in a way the proxy should surface
    to the client (e.g. image missing, healthcheck timeout)."""


# ── Orchestrator ───────────────────────────────────────────────────────
#
# The llama-server image's ENTRYPOINT (llama-server-entrypoint.sh inside
# the image) reads MODEL_FILE / MODEL_REPO / CONTEXT_SIZE / etc. from the
# container environment and assembles the llama-server command. The
# orchestrator only needs to set those env vars and start the container —
# no inline bash command, no entrypoint override.


class Orchestrator:
    """High-level Docker operations for GPU service lifecycle.

    `volumes` is the host-path registry loaded from volumes.toml. Spawn
    methods reference volumes by logical name (e.g. `llmc-llama-models`);
    the orchestrator resolves each name → host directory before passing
    the bind dict to the Docker SDK. The Docker daemon binds host paths
    directly into the new container (no Docker named-volume indirection).
    """

    def __init__(
        self,
        client=None,
        network: str = DEFAULT_NETWORK,
        volumes: Optional[VolumeRegistry] = None,
    ):
        if client is None:
            import docker
            client = docker.from_env()
        self.client = client
        self.network = network
        self.volumes = volumes  # May be None for legacy callers / tests

    # ── Volume name → host path resolution ─────────────────────────────

    def _resolve_volumes(
        self, mounts: dict[str, dict]
    ) -> dict[str, dict]:
        """Rewrite a `{logical_name: {bind, mode}}` dict into the host-path
        form Docker expects for direct bind mounts: `{/host/path: {bind, mode}}`.

        Without a registry the input is passed through unchanged so existing
        tests that mock the SDK and inspect volume names keep working.
        """
        if self.volumes is None:
            return mounts
        resolved: dict[str, dict] = {}
        for name, spec in mounts.items():
            # Already a host path (starts with /) — pass through.
            if name.startswith("/"):
                resolved[name] = spec
                continue
            host = self.volumes.device_for(name)
            # Daemon doesn't auto-create bind sources for safety; ensure
            # the directory is on disk before the container references it.
            host.mkdir(parents=True, exist_ok=True)
            resolved[str(host)] = spec
        return resolved

    # ── Mode introspection ─────────────────────────────────────────────

    def current_mode(self) -> str:
        """Return the active GPU mode based on running containers labelled
        with llmc.mode. Returns 'idle' if no labelled GPU service is up."""
        for container in self._labelled_containers():
            if container.status != "running":
                continue
            mode = container.labels.get(GPU_LABEL)
            if mode in SERVICES:
                return mode
        return "idle"

    def _labelled_containers(self) -> list:
        return self.client.containers.list(
            all=True,
            filters={"label": GPU_LABEL},
        )

    # ── Lifecycle ──────────────────────────────────────────────────────

    def stop_gpu_services(self, *, timeout: int = 10) -> list[str]:
        """Stop and remove every container with the llmc.mode label.
        Returns the names of containers that were stopped."""
        from docker.errors import NotFound
        stopped = []
        for container in self._labelled_containers():
            try:
                if container.status == "running":
                    container.stop(timeout=timeout)
                container.remove(force=True)
                stopped.append(container.name)
            except NotFound:
                continue
        return stopped

    def spawn(
        self,
        service: GpuService,
        *,
        environment: Optional[dict] = None,
        volumes: Optional[dict] = None,
        ports: Optional[dict] = None,
        command=None,
        entrypoint: Optional[list[str]] = None,
        shm_size: str = "2g",
    ):
        """Generic container spawn for a GPU service. Stops any other GPU
        service first (mutual exclusion). Returns the created container.
        Caller should call `wait_healthy()` before forwarding requests.

        IMPORTANT: pass `command` as a list[str], not a bare string. The
        Docker SDK shlex-splits string commands at whitespace, which mangles
        any non-trivial shell script (sees `if [` as three tokens, etc.).
        Use `["the script"]` to keep it as one Cmd entry."""
        from docker.errors import APIError, ImageNotFound, NotFound
        from docker.types import DeviceRequest, LogConfig

        self.stop_gpu_services()

        # Defense in depth: stop_gpu_services finds containers by the
        # llmc.mode label, but if a container with our target name exists
        # without the label (e.g. created by an older proxy version, or
        # left behind by a crashed/aborted run), `containers.run` below
        # would 409 with "container name already in use". Remove any
        # name-conflict before spawning so swap stays idempotent.
        try:
            existing = self.client.containers.get(service.name)
            existing.remove(force=True)
        except NotFound:
            pass
        except APIError:
            # If we can't remove it, let the real run() call surface the
            # error with full context rather than masking it here.
            pass

        run_kwargs = dict(
            image=service.image,
            name=service.name,
            hostname=service.hostname,
            detach=True,
            network=self.network,
            environment=environment or {},
            volumes=self._resolve_volumes(volumes or {}),
            device_requests=[DeviceRequest(count=-1, capabilities=[["gpu"]])],
            shm_size=shm_size,
            restart_policy={"Name": "unless-stopped"},
            labels={SERVICE_LABEL: service.hostname, GPU_LABEL: service.mode},
            # Bound the json-file log: spawned GPU services don't inherit
            # compose's logging limits, and llama-server logs every request.
            # Without this a multi-hour unattended run grows the container
            # log without bound.
            log_config=LogConfig(
                type=LogConfig.types.JSON,
                config={"max-size": "50m", "max-file": "3"},
            ),
        )
        if ports is not None:
            run_kwargs["ports"] = ports
        if entrypoint is not None:
            run_kwargs["entrypoint"] = entrypoint
        if command is not None:
            run_kwargs["command"] = command

        try:
            return self.client.containers.run(**run_kwargs)
        except ImageNotFound as exc:
            raise OrchestratorError(
                f"image {service.image!r} not found. "
                f"Run `make build` or `make pull` to fetch it."
            ) from exc
        except APIError as exc:
            raise OrchestratorError(
                f"docker API error starting {service.name}: {exc.explanation or exc}"
            ) from exc

    def spawn_llama(self, preset: Preset):
        """Start llama-server with the given preset. Existing GPU services
        are stopped first.

        The image's ENTRYPOINT (llama-server-entrypoint.sh) reads the
        preset env vars and assembles the llama-server command line — we
        just pass the environment and let the image do its job."""
        return self.spawn(
            LLAMA_SERVICE,
            environment=preset_to_env(preset),
            volumes={
                "llmc-llama-cache": {"bind": "/root/.cache", "mode": "rw"},
                "llmc-llama-models": {"bind": "/models", "mode": "rw"},
            },
        )

    def spawn_ninfer(self, preset: Preset):
        """Start ninfer-serve with the given preset. Existing GPU services
        are stopped first.

        The artifact directory is mounted read-only: ninfer only ever reads
        its `.ninfer` file, and unlike the llama flow there is nothing to
        download into the mount at spawn time."""
        volumes = {
            "llmc-ninfer-models": {"bind": "/models", "mode": "ro"},
        }
        if preset.ninfer and preset.ninfer.request_log_jsonl:
            volumes["llmc-ninfer-logs"] = {"bind": "/logs", "mode": "rw"}
        return self.spawn(
            NINFER_SERVICE,
            command=ninfer_command(preset),
            volumes=volumes,
        )

    def spawn_llm(self, preset: Preset):
        """Start whichever engine the preset names. Callers that used to
        reach for spawn_llama should use this instead, so a ninfer preset
        cannot silently start llama.cpp."""
        if preset.engine == "ninfer":
            return self.spawn_ninfer(preset)
        return self.spawn_llama(preset)

    def spawn_comfyui(self):
        return self.spawn(
            COMFYUI_SERVICE,
            shm_size="4g",
            ports={
                # Bind 8188 to host loopback so the ComfyUI web UI is directly
                # reachable while comfyui mode is active. Proxy at /comfyui/*
                # also works, but the websocket-based live preview only works
                # via direct connection.
                "8188/tcp": ("127.0.0.1", 8188),
            },
            volumes={
                "llmc-comfyui-models": {"bind": "/app/ComfyUI/models", "mode": "rw"},
                "llmc-comfyui-output": {"bind": "/app/ComfyUI/output", "mode": "rw"},
                "llmc-comfyui-input": {"bind": "/app/ComfyUI/input", "mode": "rw"},
                "llmc-comfyui-custom-nodes": {"bind": "/app/ComfyUI/custom_nodes", "mode": "rw"},
                "llmc-comfyui-user": {"bind": "/app/ComfyUI/user", "mode": "rw"},
            },
        )

    def spawn_train(self):
        return self.spawn(
            TRAIN_SERVICE,
            shm_size="4g",
            environment={
                "TRAIN_PORT": "8787",
                "DATA_DIR": "/data",
                "CHECKPOINTS_DIR": "/checkpoints",
            },
            volumes={
                "llmc-training-data": {"bind": "/data", "mode": "rw"},
                "llmc-comfyui-models": {"bind": "/models", "mode": "ro"},
                "llmc-comfyui-loras": {"bind": "/loras", "mode": "rw"},
            },
        )

    # ── Health check ───────────────────────────────────────────────────

    def wait_healthy(
        self,
        service: GpuService,
        *,
        timeout: Optional[int] = None,
    ) -> bool:
        """Poll the service's health endpoint until 200 or timeout.
        Uses the service hostname over the user-defined network, so this
        only works from within the proxy container (or another container
        on the same network)."""
        deadline = time.monotonic() + (timeout or DEFAULT_HEALTH_TIMEOUT)
        while time.monotonic() < deadline:
            try:
                conn = http.client.HTTPConnection(
                    service.hostname, service.internal_port, timeout=3
                )
                conn.request("GET", service.health_path)
                resp = conn.getresponse()
                resp.read()
                conn.close()
                if resp.status == 200:
                    return True
            except (OSError, http.client.HTTPException):
                pass
            time.sleep(2)
        return False

    # ── Asset downloads (mmproj, template, GGUF prefetch) ──────────────

    def ensure_asset(
        self,
        assets_dir: Path,
        filename: str,
        url: str,
        *,
        kind: str = "asset",
        timeout: int = 300,
    ) -> Path:
        """Download `url` to `assets_dir/filename` if missing. Returns the
        destination path. Idempotent. The assets_dir is the host-side bind
        path of the llmc-llama-models volume (writable from the proxy
        container which mounts it)."""
        assets_dir.mkdir(parents=True, exist_ok=True)
        dest = assets_dir / filename
        if dest.exists():
            return dest

        tmp = dest.with_suffix(dest.suffix + ".tmp")
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "llmc/2.0"})
            with urllib.request.urlopen(req, timeout=timeout) as response:
                with tmp.open("wb") as f:
                    while True:
                        chunk = response.read(1 << 20)  # 1 MiB
                        if not chunk:
                            break
                        f.write(chunk)
            tmp.rename(dest)
            return dest
        except (OSError, urllib.error.URLError) as exc:
            if tmp.exists():
                tmp.unlink()
            raise OrchestratorError(
                f"failed to download {kind} {filename} from {url}: {exc}"
            ) from exc

    def ensure_preset_assets(self, preset: Preset, assets_dir: Path) -> None:
        """Download mmproj and template files for a preset if missing.
        Skipped if the preset uses an explicit `file` reference (manually-
        placed) rather than a `url`."""
        if preset.mmproj.url and preset.mmproj_filename:
            self.ensure_asset(
                assets_dir, preset.mmproj_filename, preset.mmproj.url, kind="mmproj"
            )
        if preset.template.url and preset.template_filename:
            self.ensure_asset(
                assets_dir,
                preset.template_filename,
                preset.template.url,
                kind="template",
            )

File: proxy-go/internal/proxy/orchestrator.go
// Orchestrator: GPU service lifecycle on top of the Docker client.
// Mirror of llmc/orchestrator.py - three mutually exclusive GPU services
// (llama-server, comfyui, lora-train) found via the llmc.mode label.
package proxy

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

type GpuService struct {
	Name         string // container name
	Hostname     string // hostname on the user-defined network
	Mode         string // llm | comfyui | train
	Image        string
	InternalPort int
	HealthPath   string
}

const (
	GPULabel     = "llmc.mode"
	ServiceLabel = "llmc.service"
)

var (
	LlamaService   = GpuService{Name: "llama_server", Hostname: "llama-server", Mode: "llm", Image: envOr("LLMC_LLAMA_IMAGE", "erfianugrah/llama-server:cuda12.8-sm120"), InternalPort: 8080, HealthPath: "/health"}
	ComfyUIService = GpuService{Name: "comfyui", Hostname: "comfyui", Mode: "comfyui", Image: envOr("LLMC_COMFYUI_IMAGE", "erfianugrah/comfyui:cuda12.8-sm120"), InternalPort: 8188, HealthPath: "/system_stats"}
	TrainService   = GpuService{Name: "lora_train", Hostname: "lora-train", Mode: "train", Image: envOr("LLMC_TRAIN_IMAGE", "erfianugrah/lora-train:latest"), InternalPort: 8787, HealthPath: "/health"}
)

// NinferService serves the same mode as llama.cpp: both are the LLM and the
// GPU holds one workload at a time. It is deliberately NOT in Services (which
// maps mode -> service 1:1); resolve it from the preset via LLMServiceFor.
var NinferService = GpuService{Name: "ninfer_server", Hostname: "ninfer-server", Mode: "llm", Image: envOr("LLMC_NINFER_IMAGE", "erfianugrah/ninfer:cuda13.1-sm120a-d492968"), InternalPort: 8080, HealthPath: "/health"}

// LLMServiceFor returns the container that serves this preset. Two engines
// share mode "llm", so the mode alone cannot decide - the preset does.
func LLMServiceFor(p *Preset) GpuService {
	if p != nil && p.Engine == EngineNinfer {
		return NinferService
	}
	return LlamaService
}

// NinferCommand builds the ninfer-serve argv for a ninfer preset. Unlike the
// llama image (whose ENTRYPOINT assembles a command line from environment
// variables), ninfer-serve is argv-driven, so the proxy owns flag rendering.
func NinferCommand(p *Preset) ([]string, error) {
	if p.Engine != EngineNinfer {
		return nil, &PresetError{Msg: fmt.Sprintf(
			"NinferCommand: preset %q declares engine %q", p.Name, p.Engine)}
	}
	if p.Ninfer == nil {
		return nil, &PresetError{Msg: fmt.Sprintf("NinferCommand: preset %q has no [ninfer] section", p.Name)}
	}
	n := p.Ninfer
	argv := []string{
		"ninfer-serve",
		"/models/" + p.Model.File,
		"--model-id", p.ModelID(),
		"--host", "0.0.0.0",
		"--port", strconv.Itoa(NinferService.InternalPort),
		"--max-context", strconv.Itoa(n.MaxContext),
		"--max-concurrency", strconv.Itoa(n.MaxConcurrency),
		"--kv-dtype", n.KVDtype,
	}
	if n.Spec != "" {
		argv = append(argv, "--spec", n.Spec)
	}
	if n.KVCapacity != "" {
		argv = append(argv, "--kv-capacity", n.KVCapacity)
	}
	if n.RequestLogJsonl != "" {
		argv = append(argv, "--request-log-jsonl", n.RequestLogJsonl)
	}
	// Pointer checks, not zero checks: 0 is meaningful for the slot flags
	// (the WSL2 pinned-host workaround) and must still be emitted.
	for _, f := range []struct {
		flag string
		val  *int
	}{
		{"--draft-tokens", n.DraftTokens},
		{"--host-state-slots", n.HostStateSlots},
		{"--host-kv-mib", n.HostKVMib},
		{"--device-state-slots", n.DeviceStateSlots},
		{"--default-thinking-budget", n.DefaultThinkingBudget},
		{"--prefill-chunk", n.PrefillChunk},
		{"--max-pending-requests", n.MaxPendingRequests},
		{"--pending-timeout-ms", n.PendingTimeoutMs},
	} {
		if f.val != nil {
			argv = append(argv, f.flag, strconv.Itoa(*f.val))
		}
	}
	for _, f := range []struct {
		flag string
		on   bool
	}{
		{"--lm-head-draft", n.LMHeadDraft},
		{"--preserve-thinking", n.PreserveThinking},
		{"--vision", n.Vision},
	} {
		if f.on {
			argv = append(argv, f.flag)
		}
	}
	return argv, nil
}

var Services = map[string]GpuService{
	"llm":     LlamaService,
	"comfyui": ComfyUIService,
	"train":   TrainService,
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// OrchestratorError surfaces Docker failures to the client (503s).
type OrchestratorError struct{ Msg string }

func (e *OrchestratorError) Error() string { return e.Msg }

// Orchestrator is the scheduler's port to Docker. An interface so scheduler
// tests drive a fake (parity with Python's MagicMock orchestrator tests).
type Orchestrator interface {
	CurrentMode() string
	SpawnLlama(p *Preset) error
	// SpawnLLM dispatches on the preset's engine. Prefer it over SpawnLlama:
	// a ninfer preset handed to SpawnLlama would start the wrong engine.
	SpawnLLM(p *Preset) error
	SpawnNinfer(p *Preset) error
	SpawnComfyUI() error
	SpawnTrain() error
	WaitHealthy(svc GpuService, timeout time.Duration) bool
	EnsurePresetAssets(p *Preset, assetsDir string) error
}

// DockerOrchestrator is the production Orchestrator.
type DockerOrchestrator struct {
	Client  *DockerClient
	Network string
	Volumes *VolumeRegistry
}

func (o *DockerOrchestrator) CurrentMode() string {
	containers, err := o.Client.ListByLabel(GPULabel)
	if err != nil {
		return "idle"
	}
	for _, c := range containers {
		if c.State != "running" {
			continue
		}
		if mode := c.Labels[GPULabel]; Services[mode].Name != "" {
			return mode
		}
	}
	return "idle"
}

func (o *DockerOrchestrator) stopGPU() error {
	containers, err := o.Client.ListByLabel(GPULabel)
	if err != nil {
		return err
	}
	for _, c := range containers {
		if c.State == "running" {
			if err := o.Client.Stop(c.ID, 10); err != nil {
				return err
			}
		}
		if err := o.Client.Remove(c.ID); err != nil {
			return err
		}
	}
	return nil
}

func (o *DockerOrchestrator) resolveBinds(mounts map[string]BindSpec) (map[string]BindSpec, error) {
	out := map[string]BindSpec{}
	for name, spec := range mounts {
		host, err := o.Volumes.DeviceFor(name)
		if err != nil {
			return nil, err
		}
		out[host] = spec
	}
	return out, nil
}

func (o *DockerOrchestrator) spawn(svc GpuService, env map[string]string, mounts map[string]BindSpec, shmMB int64, ports map[string]string) error {
	return o.spawnCmd(svc, env, mounts, shmMB, ports, nil)
}

func (o *DockerOrchestrator) spawnCmd(svc GpuService, env map[string]string, mounts map[string]BindSpec, shmMB int64, ports map[string]string, cmd []string) error {
	if err := o.stopGPU(); err != nil {
		return &OrchestratorError{Msg: fmt.Sprintf("stopping GPU services: %v", err)}
	}
	// Defense in depth: remove any name-conflict container left without the
	// label (crashed run, older proxy) so create doesn't 409.
	if existing, err := o.Client.GetByName(svc.Name); err == nil && existing != nil {
		_ = o.Client.Remove(existing.ID)
	}
	binds, err := o.resolveBinds(mounts)
	if err != nil {
		return &OrchestratorError{Msg: err.Error()}
	}
	err = o.Client.CreateAndStart(CreateSpec{
		Image:     svc.Image,
		Name:      svc.Name,
		Hostname:  svc.Hostname,
		Env:       env,
		Binds:     binds,
		Network:   o.Network,
		Labels:    map[string]string{ServiceLabel: svc.Hostname, GPULabel: svc.Mode},
		ShmSize:   shmMB << 20,
		PortBinds: ports,
		Cmd:       cmd,
		GPU:       true,
	})
	if err != nil {
		return &OrchestratorError{Msg: fmt.Sprintf("starting %s: %v", svc.Name, err)}
	}
	return nil
}

func (o *DockerOrchestrator) SpawnLlama(p *Preset) error {
	return o.spawn(LlamaService, p.Env(), map[string]BindSpec{
		"llmc-llama-cache":  {Bind: "/root/.cache", Mode: "rw"},
		"llmc-llama-models": {Bind: "/models", Mode: "rw"},
	}, 2048, nil)
}

// SpawnNinfer starts ninfer-serve for the given preset. The artifact dir is
// mounted read-only: ninfer only reads its .ninfer file, and unlike the llama
// flow there is nothing to download into the mount at spawn time.
func (o *DockerOrchestrator) SpawnNinfer(p *Preset) error {
	argv, err := NinferCommand(p)
	if err != nil {
		return &OrchestratorError{Msg: err.Error()}
	}
	mounts := map[string]BindSpec{
		"llmc-ninfer-models": {Bind: "/models", Mode: "ro"},
	}
	if p.Ninfer.RequestLogJsonl != "" {
		mounts["llmc-ninfer-logs"] = BindSpec{Bind: "/logs", Mode: "rw"}
	}
	return o.spawnCmd(NinferService, nil, mounts, 2048, nil, argv)
}

// SpawnLLM starts whichever engine the preset names. Callers that reached for
// SpawnLlama should use this, so a ninfer preset cannot silently start
// llama.cpp with a config it does not understand.
func (o *DockerOrchestrator) SpawnLLM(p *Preset) error {
	if p != nil && p.Engine == EngineNinfer {
		return o.SpawnNinfer(p)
	}
	return o.SpawnLlama(p)
}

func (o *DockerOrchestrator) SpawnComfyUI() error {
	return o.spawn(ComfyUIService, nil, map[string]BindSpec{
		"llmc-comfyui-models":       {Bind: "/app/ComfyUI/models", Mode: "rw"},
		"llmc-comfyui-output":       {Bind: "/app/ComfyUI/output", Mode: "rw"},
		"llmc-comfyui-input":        {Bind: "/app/ComfyUI/input", Mode: "rw"},
		"llmc-comfyui-custom-nodes": {Bind: "/app/ComfyUI/custom_nodes", Mode: "rw"},
		"llmc-comfyui-user":         {Bind: "/app/ComfyUI/user", Mode: "rw"},
	}, 4096, map[string]string{"8188/tcp": "8188"})
}

func (o *DockerOrchestrator) SpawnTrain() error {
	return o.spawn(TrainService, map[string]string{
		"TRAIN_PORT":      "8787",
		"DATA_DIR":        "/data",
		"CHECKPOINTS_DIR": "/checkpoints",
	}, map[string]BindSpec{
		"llmc-training-data":  {Bind: "/data", Mode: "rw"},
		"llmc-comfyui-models": {Bind: "/models", Mode: "ro"},
		"llmc-comfyui-loras":  {Bind: "/loras", Mode: "rw"},
	}, 4096, nil)
}

// WaitHealthy polls the service health endpoint until 200 or timeout.
func (o *DockerOrchestrator) WaitHealthy(svc GpuService, timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	url := fmt.Sprintf("http://%s:%d%s", svc.Hostname, svc.InternalPort, svc.HealthPath)
	client := &http.Client{Timeout: 3 * time.Second}
	for time.Now().Before(deadline) {
		resp, err := client.Get(url)
		if err == nil {
			io.Copy(io.Discard, resp.Body)
			resp.Body.Close()
			if resp.StatusCode == 200 {
				return true
			}
		}
		time.Sleep(2 * time.Second)
	}
	return false
}

// EnsurePresetAssets downloads mmproj/template assets if missing.
func (o *DockerOrchestrator) EnsurePresetAssets(p *Preset, assetsDir string) error {
	if p.MMProj.URL != "" {
		if _, err := ensureAsset(assetsDir, p.MMProjFilename(), p.MMProj.URL); err != nil {
			return err
		}
	}
	if p.Template.URL != "" {
		if _, err := ensureAsset(assetsDir, p.TemplateFilename(), p.Template.URL); err != nil {
			return err
		}
	}
	return nil
}

// ensureAsset downloads url to assetsDir/filename if missing (tmp + rename).
func ensureAsset(assetsDir, filename, url string) (string, error) {
	if err := os.MkdirAll(assetsDir, 0o755); err != nil {
		return "", err
	}
	dest := filepath.Join(assetsDir, filename)
	if _, err := os.Stat(dest); err == nil {
		return dest, nil
	}
	tmp := dest + ".tmp"
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", "llmc/2.0")
	client := &http.Client{Timeout: 300 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", &OrchestratorError{Msg: fmt.Sprintf("downloading %s: %v", filename, err)}
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", &OrchestratorError{Msg: fmt.Sprintf("downloading %s: HTTP %d", filename, resp.StatusCode)}
	}
	f, err := os.Create(tmp)
	if err != nil {
		return "", err
	}
	if _, err := io.Copy(f, resp.Body); err != nil {
		f.Close()
		os.Remove(tmp)
		return "", &OrchestratorError{Msg: fmt.Sprintf("downloading %s: %v", filename, err)}
	}
	f.Close()
	if err := os.Rename(tmp, dest); err != nil {
		return "", err
	}
	return dest, nil
}

// LoadedLlamaModel returns the preset name of the model llama-server
// currently has loaded ("" if unknown/unreachable). Used at startup to fill
// state.model after a state-loss restart, avoiding a pointless swap.
func (o *DockerOrchestrator) LoadedLlamaModel(presets *PresetStore) string {
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(fmt.Sprintf("http://%s:%d/v1/models", LlamaService.Hostname, LlamaService.InternalPort))
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	var body struct {
		Data []struct {
			ID string `json:"id"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil || len(body.Data) == 0 {
		return ""
	}
	// llama-server reports the GGUF path (e.g. /models/Qwen3.8-27B-Q4_K_M.gguf)
	base := strings.TrimSuffix(filepath.Base(body.Data[0].ID), ".gguf")
	if p := presets.ByName(base); p != nil {
		return p.Name
	}
	return ""
}

// ValidMode reports whether name is a switchable GPU mode.
func ValidMode(name string) bool {
	_, ok := Services[name]
	return ok
}

// ServiceFor returns the GpuService for a mode.
func ServiceFor(mode string) (GpuService, bool) {
	s, ok := Services[mode]
	return s, ok
}
